processor/logdeduplicationprocessor/processor.go

```diff
@@ -0,0 +1,133 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package logdeduplicationprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdeduplicationprocessor"
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+// logDedupProcessor is a logDedupProcessor that counts duplicate instances of logs.
+type logDedupProcessor struct {
+	emitInterval time.Duration
+	aggregator   *logAggregator
+	remover      *fieldRemover
+	consumer     consumer.Logs
+	logger       *zap.Logger
+	cancel       context.CancelFunc
+	wg           sync.WaitGroup
+	mux          sync.Mutex
+}
+
+func newProcessor(cfg *Config, consumer consumer.Logs, logger *zap.Logger) (*logDedupProcessor, error) {
+	// This should not happen due to config validation but we check anyways.
+	timezone, err := time.LoadLocation(cfg.Timezone)
+	if err != nil {
+		return nil, fmt.Errorf("invalid timezone: %w", err)
+	}
+
+	return &logDedupProcessor{
+		emitInterval: cfg.Interval,
+		aggregator:   newLogAggregator(cfg.LogCountAttribute, timezone),
+		remover:      newFieldRemover(cfg.ExcludeFields),
+		consumer:     consumer,
+		logger:       logger,
+	}, nil
+}
+
+// Start starts the processor.
+func (p *logDedupProcessor) Start(_ context.Context, _ component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	p.cancel = cancel
+
+	p.wg.Add(1)
+	go p.handleExportInterval(ctx)
+
+	return nil
+}
+
+// Capabilities returns the consumer's capabilities.
+func (p *logDedupProcessor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
+}
+
+// Shutdown stops the processor.
+func (p *logDedupProcessor) Shutdown(ctx context.Context) error {
+	if p.cancel != nil {
+		p.cancel()
+	}
+
+	doneChan := make(chan struct{})
+	go func() {
+		defer close(doneChan)
+		p.wg.Wait()
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-doneChan:
+		return nil
+	}
```

Comment:
Is all of this extraneous? Create a local done channel, listen for it to close, but defer closing it?

The only thing somewhat useful this might be doing is returning `ctx.Err()` in the case where the passed in context is canceled while we're shutting down. I'm not convinced that's useful though and not aware of any other cases where we do that in `Shutdown`.

Reply:
Opened tracking issue:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34478